lib/src/gitattributes.rs

```diff
@@ -0,0 +1,193 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
+
+use gix::attrs as gix_attrs;
+use gix::glob as gix_glob;
+use gix::path as gix_path;
+use std::borrow::Cow;
+use std::io;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum GitAttributesError {
+    #[error("Failed to read attributes patterns from file {path}")]
+    ReadFile { path: PathBuf, source: io::Error },
+}
+
+/// Models the effective contents of multiple .gitattributes files.
+#[derive(Debug)]
+pub struct GitAttributesFile {
+    search: gix_attrs::Search,
+    collection: gix_attrs::search::MetadataCollection,
+}
+
+impl GitAttributesFile {
+    pub fn chain(
+        self: &Arc<GitAttributesFile>,
+        prefix: PathBuf,
+        input: &[u8],
+    ) -> Result<Arc<GitAttributesFile>, GitAttributesError> {
+        let mut source_file = prefix.clone();
+        source_file.push(".gitattributes");
+
+        let mut search = self.search.clone();
+        let mut collection = self.collection.clone();
+        search.add_patterns_buffer(input, source_file, Some(&prefix), &mut collection, true);
+        Ok(Arc::new(GitAttributesFile { search, collection }))
+    }
+
+    /// Concatenates new `.gitattributes` file.
+    ///
+    /// The `prefix` should be a slash-separated path relative to the workspace
+    /// root.
+    pub fn chain_with_file(
+        self: &Arc<GitAttributesFile>,
+        prefix: &str,
+        file: PathBuf,
+    ) -> Result<Arc<GitAttributesFile>, GitAttributesError> {
+        if file.is_file() {
+            let mut buf = Vec::new();
+            let mut search = self.search.clone();
+            let mut collection = self.collection.clone();
+            search
+                .add_patterns_file(
+                    file.clone(),
+                    true,
+                    Some(Path::new(prefix)),
+                    &mut buf,
+                    &mut collection,
+                    true,
+                )
+                .map_err(|err| GitAttributesError::ReadFile {
+                    path: file.clone(),
+                    source: err,
+                })?;
+            Ok(Arc::new(GitAttributesFile { search, collection }))
+        } else {
+            Ok(self.clone())
+        }
+    }
+
+    pub fn matches(&self, path: &str) -> bool {
+        // If path ends with slash, consider it as a directory.
+        let (path, is_dir) = match path.strip_suffix('/') {
+            Some(path) => (path, true),
+            None => (path, false),
+        };
+
+        let mut out = gix_attrs::search::Outcome::default();
+        out.initialize_with_selection(&self.collection, ["filter"]);
+        self.search.pattern_matching_relative_path(
+            path.into(),
+            gix_glob::pattern::Case::Sensitive,
+            Some(is_dir),
+            &mut out,
+        );
```

Comment:
I've been trying to use this branch with my LFS setup and it wasn't working.

It turns out that my `.gitattributes` in root was:
```
path/to/data/*.bin filter=lfs diff=lfs merge=lfs -text
```
This code only works fore relative path, so I had to move my .gitattributes to `path/to/data/.gitattributes` and use relative lfs.

Reply:
I took a quick look and curiously, it seems like relatively pathed files do match the search, but wind up with an assignment with an ["unspecified"](https://docs.rs/gix-attributes/0.24.0/gix_attributes/enum.State.html#variant.Unspecified) filter value (instead of the expected `Value("lfs")`). Not sure if I'm using the API incorrectly or if there's perhaps an upstream bug.